calref/cboe
# Record advance time

Fix #437

* Record and replay actions from #422
* Implement `--verbose` flag for replay recording, which records internal testing actions (for now just `advance_time`)
* This PR I think covers making the vast majority of action cases perform advance_time() identically to original runtime behavior. I would say it's highly likely that I missed edge cases, but this version of the code is much more correct.
* Also of note: obsessively matching the behavior of `advance_time()` while correct and good, is not entirely necessary for replay integrity as long as action replays always handle the `did_something` flag correctly.

To test this and find every discrepancy, I used these two replays which together contain one of every action (I'm pretty sure):

[Archive.zip](https://github.com/user-attachments/files/17137020/Archive.zip)

There are some major caveats to those 2 replays, which is why I am not staging them in the repository:

* They were both recorded with different builds, each in a modified state
* In the first one, I deleted some actions at the end which led to the crash #455
* As I was testing and fixing, I changed the way some actions record themselves. Rather than re-record the whole logs, I would edit the log to match the changing behavior (but I could have made a mistake).
* They take very very long.

Some things about these replays are cool:
* They recreate pretty long play sessions, appearing to match the state of the original playthrough perfectly, including in combat (which is fun to watch)
* You can see the replays reproduce some UI bugs which would be impossible to check with unit testing
* If I ever demo OpenBoE at a public event, I'll probably make replays to have looping when no one's actively playing it :)

This is a wibbly wobbly PR which I'm afraid rebasing for cleaner history, would get too hairy. So I haven't tried squashing anything yet. I also don't think it would be proper to squash the whole PR into one commit, because it shows the evolution of the `--verbose` flag and my reasoning for it. All the contiguous "fix some cases" commits could probably be squashed together easily, though.

My next recording system PRs after this one will do the following to address the dangling problems:

* Add two folders, `test/replays/short` and `test/replays/long`
* Add a much shorter log that records one of everything FROM A CLEAN REPO STATE, WITHOUT MANUAL MODIFICATIONS
* Add a bash script `test/replays.sh` which runs the replays, with the options to fail fast and/or skip the long ones.
* On success of a comprehensive run of all replays, this script will update a file `test/replays/success_{platform}.txt` which records the latest gitrev info unless the repository has been modified. This file can be staged, and potentially the CI for release builds could check for its existence and assert that a pull request comes with a recent success SHA for all 3 platforms (though this requires a dev to have access to all of Mac/Windows/Linux, which I do for now, but obviously not everyone will and my computers might break some day.)
* Get the replay test to pass on all 3 platforms so that `success` files can exist for all 3.
